lsp/nls/src/field_walker.rs

```diff
@@ -34,6 +34,42 @@ impl From<RichTerm> for TermAtPath {
     }
 }
 
+/// The things that can appear as the right hand side of a `Def`
+#[derive(Clone, Debug, PartialEq)]
+pub enum Value {
+    /// In `let foo = value`, the term `value` is the right hand side of the `Def`.
+    Term(RichTerm),
+    /// If type inference has determined that `x` is of type
+    ///
+    /// `{ foo: Number }`
+    ///     ^-----------------------------------------------------------|
+    /// then in the expression `x.foo`, the definition of `foo` is here | and the
+    /// value of its definition is `Number`.
```

Comment:
The idea here is that `Value` (maybe not well named) is a thing that might have fields for recursing into when resolving paths. It could be a term (like `{ foo = 1 }`) or a type (like `{ foo : Number }`). For completion (and eventually goto definition) we need to be able to recurse through both of these cases. So for example, if `x` has type `{ foo : { bar : { baz : Number }}}` and we ask for the fields on `x.foo.bar`, then in the first call to `FieldDefs::resolve` we'll find that `x` has a field `foo` with "value" `{ bar : { baz : Number }}`, and then on the next call to `FieldDefs::resolve` we'll find the subfield `bar` with "value" `{ baz: Number }`.

The recursion can look at both terms and types/contracts. For example, if you have
```let x | { bar, baz, .. } = { blorg } in x.```
then in the `Term::Annotated` branch we collect the completions from both the annotation and the contract.  We also accumulate fields from inferred types in `FieldDefs::resolve`, and combine those fields with the ones from the term.

Thanks to this question, I realized that I'd forgotten to accumulate fields from contracts and types in record field metadata; this is fixed in the latest commit. With the fix, if you call `resolve_path` on
```
{ foo : { bar : Number } = { bar = 1 }}
```
with the path `foo`:
- on the first iteration, we'll see that the overall expression has an ident `foo`, and its `Def` has value `{ bar = 1 }` and metadata `{ bar : Number }`.
- on the second iteration, we'll recurse into the previous `Def` for `foo`. We'll visit the value, which has a field `bar` (whose Def has value 1, and no metadata). And we'll visit the metadata, which has a field `bar` (whose Def has value `Number` and no metadata). So we'll return two `Def`s, one for each occurrence of `bar`.

Reply:
Sorry, I used definition/Defn a lot, but I believe I'm rather talking about `Value` : each field should have one definition, but instead of having one value, it would make more sense to me that they have multiple completion sources. Or, if you don't want to duplicate metadata and mimic the original's field structure, keep the metadata separate and set value to `None` for `{foo : {bar : Number}}`. I think what's confusing me is that we do something in between: when there's a definite value we keep the same structure as the original field - value inside `value` and the rest in the metadata -, but for some reason, when there's no definition, we put the type annotation in the value?

Reply:
Ok thanks, I'll see if I can find a better model for this. A few considerations:
- the name `Def` was chosen because the same infrastructure is supposed to eventually drive "goto definition/goto declaration"
- we want to keep (at least part of) the metadata attached to fields: it's used for generating documentation, and not just for finding subfields

Maybe not relevant for much longer, but the current determinant for whether `Value` is a type or a term is that when visiting a type in `resolve_type`, the value is a type. When visiting a term in `resolve_term`, the value is a term.

Reply:
Ok, I think I better see your point of "when resolving a path, we might end up on a subpart either of a type of a term" now. Then do you use metadata only for the documentation? Can `Def.metadata` be set if `value` is a type for example?

During resolution, what happens if you end up on a `value` that has both a value definition, contract annotations and type annotations (say at on a field such as `{foo | {..1..} : {..2..} = {..3..}`. Does it gives rise to as many other values that you explore (one for each item), with the path copied?

Sorry for all the questions :sweat_smile:

Reply:
Not at all, questions are great!

For your first question, we're currently using it only for documentation. There are potentially other metadata fields in the completion response that we could be filling out; I haven't looked in much detail yet. Right now `metadata` is only set for fields in a `Term::Record`, so I guess that means it's only paired with value being a `term`.

The answer depends on what "end up on a value" means:
- if you ask for all the fields on `{foo | {..1..} : {..2..} = {..3..}}` then you will get back a single `Def`, with identifier `foo`, value `{..3..}`, and the rest in `metadata`
- but if you take `{foo | {..1..} : {..2..} = {..3..}}` and ask "what are the fields on the `foo` field?" (i.e. you call `resolve_path` with path `[foo]`) then there will be three recursive calls to `FieldDefs::resolve`: one for the value, one for the contract, and one for the type. The results of those will be merged together.

Reply:
Ok, I think this modelling issue might be resolved by a refactor that I was thinking of doing anyway: making `resolve` one step lazier than it currently is. That is, currently `resolve` returns a collection of `Ident`s and `Def`. Instead, I want `resolve` to return a collection of things that *have* fields. So basically, the old way can be obtained from the new way by flat-mapping over the things that have fields and getting all their fields.

"Things that have fields" are
```rust
#[derive(Clone, Debug, PartialEq)]
pub enum FieldHaver { // I'd welcome a better name suggestion!
    RecordTerm(RecordData),
    Dict(Type),
    RecordType(RecordRows),
}
```

There are three advantages of this:
- efficiency: when resolving a long path, we aren't interested in most fields and so it was wasteful to collect them all
- avoids some special-casing to handle the annoyance that we *can't* collect all fields of a `Dict`
- avoids the need to have a `Def` type describing the "value" of a field.

I have a draft of this and it seems to work ok. If it sounds reasonable, I'll finish it up on Monday.

Reply:
FWIW, from a conceptual point of view, I do find the `FieldHaver` modeling/naming idea simpler and more natural, indeed.